README
B::Deobfuscate version 0.17
==========================

B::Deobfuscate is a backend module for the Perl compiler that
generates perl source code, based on the internal compiled structure
that perl itself creates after parsing a program. It adds symbol
renaming functions to the B::Deparse module. An obfuscated program is
already parsed and interpreted correctly by the B::Deparse
program. Unfortunately, if the obfuscation involved variable renaming
then the resulting program also has obfuscated symbols.

This module takes the last step and fixes names like $z5223ed336 to be
a word from a dictionary. While the name still isn't meaningful it is
at least easier to distinguish and read. Here are two examples - one
from B::Deparse and one from B::Deobfuscate.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

 B::Keywords
 YAML

COPYRIGHT AND LICENCE

Copyright (C) 2007 Joshua b. Jore. All rights reserved.

This program is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; version 2.


Changes
--- %YAML:1.2
0.01:
  - >-
    original version; created by h2xs 1.5 with options -X -P -n
    B::Deobfuscate
0.02:
  - Dictionary sort is now predictable
  - Added config file support
0.03:
  - Fixed Build.PL
0.04:
  - >-
    Fixed a bug in gv_should_be_renamed where number variables higher
    than 9 wouldn't be recognized.
  - Added sigil detection to rename_gv
  - Added Makefile.PL wrapper for Build.PL
  - Use only symbols that are four or more characters
0.05: ~
0.06:
  - Added an overridable *B::Deobfuscate::DATA dictionary
  - Added an internal dictionary
  - >-
    Support 5.6.1, 5.8.0 and bleadperl circa July 5, 2003 (which is nearly
    5.8.1)
0.07:
  - Added some sanity checking so that writing extensions is easier
0.08:
  - Removed a v-string
0.09:
  - >-
    Removed a stupid and unrelated test that always failed. I'll create proper
    tests once I learn how to use Test::Harness.
0.10:
  - >-
    broquaint of perlmonks.org noticed that I forgot to handle
    the helem opcode. This attempts to handle all the cases I
    need to (after consulting opcode.pl of the perl source).
  - Added some real tests.
0.11:
  - The dictionary may now have words that are shorter than 3 characters
  - Fixed for 5.10
0.12:
  - >-
    Updated the signature cuz the previous version forgot to account
    for the change to META.yml
  - "*foo{CODE} -> defined( &foo ) per the docs."
0.13:
  - uh... broke stuff? The SIGNATURE is bad for META.yml anyway.
  - I forget what else changed
0.14:
  - Shuffled the PGP dictionary around and introduced a floral one
  - Fussing
  - autouse more of YAML
  - autouse Carp
  - Preliminary work for renaming classes & methods
0.15:
  - Removed SIGNATURE
  - Fixed \ in $^X in t/11deparse.t
  - Perl::Critic fussing.
  - Disable broken debugging code.
  - Require latest B::Keywords. Exports were changed.
0.16:
  - Enabled alternate dictionary modules
  - Added -D parameter to ->new
0.17:
  - Removed some dups from Flowers.pm
  - Made this Changes file into YAML for kicks.
0.18:
  - Fixed t/canon.t to be 5.6.x compatible.
  - Quieter B imports.


lib/B/Deobfuscate/Dict/Flowers.pm
package B::Deobfuscate::Dict::Flowers;

=head1 NAME

B::Deobfuscate::Dict::Flowers

=cut

use strict;
use warnings;

chomp( my @words = <DATA> );
close DATA;

for (@words) {
    tr/'//d;
    s/[^-\w\n]+/_/g;
    tr/_//s;
    s/_([a-z])/\u$1/ig;
}

## no critic
no strict 'refs';
${ +__PACKAGE__ } = join '', map "$_\n", @words;

1;

__DATA__
Alp Lily
Alpine Hulsea
Alpine Sunflower
Alpine Saxifrage
Alpine Sorrel
Alpine Spiraea
American Speedwell
American Twinflower
Annual Phlox
Antelope Bitterbrush
Arctic Gentian
Arrow-leaved Balsamroot
Arrow-leaved Thelypody
Arthur's Milkvetch
Bachelor's Button
Back's Sedge
Ballhead Waterleaf
Bear Grass
Bearberry
Beautiful Shooting Star
Big-Headed Clover
Big-leaved Huckleberry
Big-Podded Mariposa Lily
Big-Podded Mariposa Lily
Biscuit Root
Bitter Cherry
Bitterroot Lewisia
Bittersweet
Black Henbane
Black Knapweed
Black Raspberry
Black Twinberry
Bearberry Honeysuckle
Blanket Flower
Blue Cup
Blue Mountain Loco Weed
Blue Mountain Onion
Blue Mountain Penstemon
Blue Mountain Swamp Onion
Blue Vine Clematis
Blue-Eyed Grass
Bog Buckbean
Bog St. John's Wort
Bracted Lousewort
Brewer's Cliff Brake
Brewer's Monkeyflower
Broom Buckwheat
Buckbrush
Sticky Laurel
Buffalo Bur
Burke's Larkspur
Butter and Egs
Calico Flower
California Fire Chalice
California Fire Chalice
California
Northern Buckwheat
California Pitcher Plant
California Poppy
Calypso Orchid
Moccasin Flower
Camas Lily
Blue Camas
Canada Thistle
Carey's Balsamroot
Cascades Douglasia
Wenatchee Douglasia
Cat's Ear Lily
Centaur Flower
Dwarf Hesperochiron
Clasping Pepper Grass
Clasping-leaved Twisted Stalk
Clustered Lady's Slipper
Coast Fawnlily
Colorado Blue Columbine
Columbia Puccoon
Wayside Gromwell
Columbian Coreopsis
Columbian Monkshood
Common Cattail
Common Grapefern
Least Moonwort
Common Prince's Pine
Common Yellow Monkeyflower
Cous
Cow Parsnip
Crenulate Moonwort
Cup Clover
Curl-leaf Mountain Mahogany
Cusick's Monkeyflower
Cut-leaved Fleabane
Dagger Pod Mustard
Dalmatian Toad Flax
Dandelion
Dark Woods
Round-leaved Violet
Davidson's Penstemon
Desert Lily
Desert Willow
Desolation Meadow Grapefern
Diffuse Knapweed
Douglas' Clover
Douglas Fir
Douglas' Onion
Drummond's Anemone
Dusty Maiden
Dougla' Chaenactis
Dutchman's Breeches
Dwarf Cornel
Bunchberry
Dwarf Monkeyflower
Early Blue Violet
Elegant Death Camas
Elk Weed
Monument Plant
Engelmann Spruce
Fairybells
False Hyacinth
Douglas' Brodiaea
Fan-leaved Cinquefoil
Fireweed
Willow Herb
Foxfire
Skyrocket Gilia
Fringe Pod
Giant Helleborine Orchid
Glacier Lily
Fawn Lily
Gold Fields
Golden Canbya
Golden Fleabane
Golden-fruited Sedge
Grand Fir
Grass-of-Parnassus
Great Blazingstar
Green False Hellebore
Green
Tall Phacelia
Hairy Owl Clover
Harebell
Harsh Indian Paintbrush
Heart-leaved Arnica
Hiker's Gentian
Hound's Tongue
Howell Dimersia
Hudson's Bay Currant
Indian Pipe
Inflated Sedge
James' Saxifrage
John Day Valley Desert Parsley
Kern Daisy
Kinniknnick
Klamath Weed
Lance-leaved Grapefern
Lance-leaved Spring Beauty
Lance-leaved Stonecrop
Large Mountain Monkeyflower
Large-flowered Collomia
Large-flowered Tonella
Leafy Spurge
Leopard Lily
Chocolate Lily
Lewis' Mockorange
Lewis' Monkeyflower
Little Prince's Pine
Lodgepole Pine
Long-Bearded Sego
Mariposa Lily
Long-flowered Bluebells
Low
Creeping Oregon Grape
Macdougal's Pincushion
Macfarlane's Four O'Clock
Maguire Lewisia
Maidenhair Fern
Male Fern
Mallow Ninebark
Marsh Marigold
Elkslip
Mary Blue-Lips
Blue-eyed Mary
Matrimony Vine
Meadow Pussytoes
Merten's Mountain Heather
Miner's Lettuce
Mingan Grapefern
Morning Glory
Moss Campion
Moss Gentian
Moth Mullein
Mountain Ash
Mountain Buttercup
Mountain Dryas
White Dryas
Mountain Grapefern
Mountain Heather
Mountain Lady's Slipper
Mountain Mariposa
Mountain Monardella
Mountain Pea
Mountain Thermopsis
Munro's Scarlet Globemallow
Musk Monkeyflower
Naked-stemmed Desert Parsley
Narrlow-leaved Indian Lettuce
Narrow-leaved Collomia
Nettle-leaved Horsemint
Nevada Primrose
Nootka Rose
Spaulding's Rose
Northern Fairy Candelabra
Northern Mule's Ears
Northern Starflower
Northern Sweet Vetch
Nuttall's Larkspur
Nuttall's Linanthastrum
Ocean Spray
One-flowered Broomrape
One-flowered Gentian
Orange Agoseris
Orange Honeysuckle
Orange Jewelweed
Orange Balsam
Orcutt's Brodiaea
Oregon Bolandra
Oregon Bottle Gentian
Oregon Boxwood
Oregon Lily
Oregon Sunshine
Oregon Wild Cucumber
Manroot
Pacific Azelea
Pacific Dogwood
Pacific Rhododendron
Pacific Yew
Parry's Primrose
Pennell's Penstemon
Pepperpod
Phantom Orchid
Phantom Orchid
Pheasant's Eye
Pine Broomrape
Pinedrops
Pink Elephantheads
Pink Fawnlily
Pink Pinwheels
Dusky Horkelia
Pinnate Grapefern
Pioneer Violet
Johnny Jump Up
Piper's Windflower
Plantain-leaved Buttercup
Plumed Clover
Pussy Clover
Porcupine Sedge
Prairie Lupine
Prairie Smoke
Old-Man's Whiskers
Prickly Pear Cactus
Prickly Phlox
Prickly Poppy
Puncture Vine
Goat Heads
Purple Loosestrife
Purple-Eyed Grass
Grass Widow
Queencup Beadlily
Ragged Robin
Dearhorn Clarkia
Ramshaw Sandverbena
Rattlesnake Brome
Red Elderberry
Red Kittentail
Red Osier Dogwood
Rock Penstemon
Rocky Mountain Beeplant
Rosy Balsamroot
Ruby Mountains Primrose
Sabin's Lupine
Sacramento Mountains Prickly Poppy
Sagebrush Buttercup
Salal
Salt Heliotrope
Sand Lily
Scabland Blepharipappus
Scalepod
Scotch Bellflower
Scotch Broom
Scotch Thistle
Scot's Broom
Sego Lily
Self-Heal
Serrated-leaved Balsamroot
Service Berry
Sarvis
Sheep Sorrel
Sour Grass
Shining
Cascades Oregon Grape
Shooting Star
Short-flowoered Monkeyflower
Showy Milkweed
Sickle-leaved Lousewort
Sierra Pea
Sierran Onion
Sierran Spring Beauty
Silky Phacelia
Simpson's Ball Cactus
Single-leaf Pinyon Pine
Siskiyou Fireweed
Skunk Cabbage
Sky Pilot
Alpine Skunkbush
Sky Pilot
Jacob's Ladder
Sleepy Cat
Slinkpod
Oregon Saxifrage
Small-flowered Fringecup
Smooth Blazingstar
Snapdragon Skullcap
Snow Willow
Snowy Spring Parsley
Spalding's Silene
Spalding's Catchfly
Spotted Coralroot Orchid
Spring Whitlow Grass
Steer's Head
Stemless Evening Primrose
Stickly Geranium
Sticky Currant
Sticky Penstemon
Sticky Phlox
Streambank Saxifrage
Brook Saxifrage
Sugar Bowls
Leather Flower
Sulfur Buckwheat
Sulfur Cinquefoil
Swale Desert Parsley
Swamp Saxifrage
Tailcup Lupine
Tall Mountain Bluebells
Tansy Ragwort
Tansy-leaved Evening Primrose
Teasel
Thimbleberry
Thin-leaved Owl Clover
Thin-leaved Paintbrush
Three-leaf Lewisia
Tidytips
Tiny Vetch
Tufted Evening Primrose
Tufted
Douglas' Phlox
Two-spiked Moonwort
Umatilla Gooseberry
Umbellate Spring Beauty
Utah Honeysuckle
Venus' Looking Glass
Wakas
Indian Pond Lily
Wake Robin
Red Trillium
Wallowa Mountains
Cusick's Primrose
Washington Lily
Washington Monkeyflower
Water Cress
Wax Currant
Wenaha Currant
Western Bleedingheart
Western Blue Flax
Western Burnet
Western Choke Cherry
Western Columbine
Western Larch
Western Lily
Western Pasqueflower
Western Redbud
Western White Pine
Western Yarrow
Western Yellow Pine
Weston's Mariposa Lily
White Alder
White Mountain Azalea
White Trillium
Whitetop
Whitney's Loco Weed
Wild Ginger
Wild Iris
Western Blue Flag
Wild Paeony
Wild Strawberry
Wintergreen
Wood Buttercup
Wood Lily
Rocky Mountain Lily
Wood Nymph
Single Delight
Woodland Beardtongue
Woolly Balsamroot
Woolly Breeches
Wyoming Paintbrush
Yellow Bell
Yellow Star Thistle
Yellow Toadflax


lib/B/Deobfuscate.pm
package B::Deobfuscate;
use strict;
use warnings;

use B qw( main_cv main_root main_start );

BEGIN {
    for my $func (qw( begin_av init_av check_av end_av )) {

        ## no critic
        no strict 'refs';
        if ( defined &{"B::$func"} ) {
            B->import($func);
        }
        else {

           # If I couldn't create it, I'll just declare it to keep lint happy.
            eval "sub $func;";
        }
    }

    # B::perlstring was added in 5.8.0
    if ( defined &B::perlstring ) {
        B->import('perlstring');
    }
    else {
        *perlstring = sub { '"' . quotemeta( shift @_ ) . '"' };
    }
}
use B::Keywords qw( @Barewords @Symbols );

use base 'B::Deparse';

use autouse Carp => 'confess';
use IO::Handle ();
use autouse YAML => qw( LoadFile Dump );

# use Data::Postponed 'postpone_forever';
sub postpone_forever { return shift @_ }

our $VERSION = '0.18';

sub load_keywords {
    my $self = shift @_;
    my $p    = $self->{ +__PACKAGE__ };

    return $p->{keywords} = {
        map { $_, undef } @Barewords,

        # Snip the sigils.
        map { substr $_, 1 } @Symbols
    };
}

sub load_unknown_dict {
    my $self = shift @_;
    my $p    = $self->{ +__PACKAGE__ };

    my $dict_data;

    # slurp the entire dictionary at once
    if ( defined( my $dict_file = $p->{unknown_dict_file} ) ) {
        open my $fh, '<', $dict_file
            or confess "Cannot open dictionary $dict_file: $!";
        local $/;    ## no critic
        $dict_data = [<$fh>];
    }
    else {
    LOAD_DICTIONARY_MODULE:
        for my $module ( $p->{unknown_dict_module}, 'PGPHashKeywords',
            'Flowers' )
        {
            next if not defined $module;
            eval "require B::Deobfuscate::Dict::$module";    ## no critic
            next if $@;

            no strict 'refs';                                ## no critic
            $dict_data = ${"B::Deobfuscate::Dict::$module"};
            last LOAD_DICTIONARY_MODULE;
        }
    }

    unless ($dict_data) {
        confess "The symbol dictionary was empty!";
    }

    my $k = $self->load_keywords;

    $p->{unknown_dict_data} = [
        sort { length $a <=> length $b or $a cmp $b }
            grep { not( /\W/mx or exists $k->{$_} ) }
            split /\n/mx,
        $dict_data
    ];

    unless ( scalar @{ $p->{'unknown_dict_data'} } ) {
        confess "The symbol dictionary is empty!";
    }

    return;
}

sub next_short_dict_symbol {
    my $self = shift @_;
    my $p    = $self->{ +__PACKAGE__ };

    my $sym = shift @{ $p->{unknown_dict_data} };
    push @{ $p->{used_symbols} }, $sym;

    unless ($sym) {
        confess "The symbol dictionary has run out and is now empty";
    }

    return $sym;
}

sub next_long_dict_symbol {
    my $self = shift @_;
    my $p    = $self->{ +__PACKAGE__ };

    my $sym = pop @{ $p->{unknown_dict_data} };
    push @{ $p->{used_symbols} }, $sym;

    unless ($sym) {
        confess "The symbol dictionary has run out and is now empty";
    }

    return $sym;
}

sub load_user_config {
    my $self        = shift @_;
    my $p           = $self->{ +__PACKAGE__ };
    my $config_file = $p->{user_config};

    return unless $config_file;

    unless ( -f $config_file ) {
        confess "Configuration file $config_file doesn't exist";
    }

    my $config = ( LoadFile($config_file) )[0];
    $p->{globals_to_ignore} = $config->{globals_to_ignore};
    $p->{pad_symbols}       = $config->{lexicals};
    $p->{gv_symbols}        = $config->{globals};
    if ( $config->{dictionary} ) {
        $p->{unknown_dict_file} = $config->{dictionary};
    }
    if ( $config->{global_regex} ) {
        $p->{global_regex} = qr/$config->{global_regex}/;
    }

    # Symbols that are listed with an undef value actually
    # just aren't renamed at all.
    for my $symt_nym (qw/pad gv/) {
        my $symt = $p->{ $symt_nym . "_symbols" };
        for my $symt_key ( keys %$symt ) {
            if ( not defined $symt->{$symt_key} ) {
                $symt->{$symt_key} = $symt_key;
            }
        }
    }

    return;
}

sub gv_should_be_renamed {
    my ( $self, $sigil, $name ) = @_;
    my $p = $self->{ +__PACKAGE__ };
    my $k = $p->{keywords};

    confess("Undefined sigil") unless defined $sigil;
    confess("Undefined name")  unless defined $name;

    # Ignore keywords
    return
        if exists $k->{$name}
        or "$sigil$name" =~ m{\A\$\d+\z}mx;

    if ( exists $p->{gv_symbols}{$name}
        or $name =~ $p->{gv_match} )
    {
        return 1;
    }
    return;
}

sub rename_pad {
    my ( $self, $name ) = @_;
    my $p = $self->{ +__PACKAGE__ };

    my ($sigil) = $name =~ m{\A(\W+)}mx
        or confess "Invalid pad variable name $name";

    my $dict = $p->{pad_symbols};
    return $dict->{$name} if $dict->{$name};

    #    $dict->{$name} = $name;
    $dict->{$name} = postpone_forever $sigil . $self->next_short_dict_symbol;

    unless ( $dict->{$name} ) {
        confess "The suggested name for the lexical variable $name is empty";
    }
    return $dict->{$name};
}

sub lookup_sigil {
    my $rv = shift @_;

    return $rv =~ /(?:gv|pad|rv2)sv\z/mx ? '$'
        : $rv =~ /(?:gvav|padav|av2arylen|rv2av|aelemfast|aelem|aslice)\z/mx
        ? '@'
        : $rv =~ /(?:padhv|rv2hv|helem|hslice)\z/mx ? '%'
        : $rv =~ /rv2cv\z/mx                        ? '&'
        : $rv =~ /(?:gv|gelem|rv2gv)\z/mx           ? ''
        :

        # Nothing valid;
        ();
}

sub rename_gv {
    my ( $self, $name ) = @_;
    my $p = $self->{ +__PACKAGE__ };

    my $sigil_debug = '';
    my $sigil;
FIND_SIGIL: {
        for ( my $cx = 0; not defined $sigil; ++$cx ) {
            my ( undef, undef, undef, $rv ) = caller $cx;
            if ( not $rv ) {
                confess
                    "No sigil could be found. Please report the following text:\n$sigil_debug\n";
            }

            $sigil = lookup_sigil($rv);

            $sigil_debug .= "$cx = $rv\n";
        }
    }

    unless ( defined $sigil ) {
        confess
            "No sigil could be found. Please report the following text:\n$sigil_debug\n";
    }

    return $name unless $self->gv_should_be_renamed( $sigil, $name );

    my $dict = $p->{gv_symbols};

    my $sname = "$sigil$name";
    return $dict->{$sname} if exists $dict->{$sname};
    $dict->{$sname} = postpone_forever $self->next_long_dict_symbol;

    unless ( $dict->{$sname} ) {
        confess "$sname could not be renamed.";
    }

    return $dict->{$sname};
}

## OVERRIDE METHODS FROM B::Deparse

sub new {
    my $class = shift @_;
    my $self  = $class->SUPER::new(@_);
    my $p     = $self->{ +__PACKAGE__ } = {};
    $p->{unknown_dict_file}   = undef;
    $p->{unknown_dict_module} = undef;
    $p->{unknown_dict_data}   = undef;
    $p->{user_config}         = undef;
    $p->{gv_match}            = qw/\A[[:lower:][:digit:]_]+\z/;
    $p->{pad_symbols}         = {};
    $p->{gv_symbols}          = {};
    $p->{output_yaml}         = 0;
    $p->{output_fh}           = \*STDOUT;

    while ( my $arg = shift @_ ) {
        ## no critic
        if ( $arg =~ m{\A-d([^,]+)} ) {
            $p->{unknown_dict_file} = $1;
        }
        elsif ( $arg =~ m{\A-D([^,]+)} ) {
            $p->{unknown_dict_module} = $1;
        }
        elsif ( $arg =~ m{\A-c([^,]+)} ) {
            $p->{user_config} = $1;
        }
        elsif ( $arg =~ m{\A-m/([^/]+)/} ) {
            $p->{gv_match} = $1;
        }
        elsif ( $arg =~ m{\A-y} ) {
            $p->{output_yaml} = 1;
        }
    }

    $self->load_user_config;
    $self->load_unknown_dict;

    return $self;
}

sub compile {    ## no critic Complex
    my (@args) = @_;

    return sub {
        my $source = '';
        my $self   = __PACKAGE__->new(@args);

        # First deparse command-line args
        if ( defined $^I ) {    # deparse -i
            $source .= q(BEGIN { $^I = ) . perlstring($^I) . qq(; }\n);
        }
        if ($^W) {              # deparse -w
            $source .= qq(BEGIN { \$^W = $^W; }\n);
        }
        ## no critic PackageVar
        if ( $/ ne "\n" or defined $O::savebackslash ) {    # deparse -l -0
            my $fs = perlstring($/) || 'undef';
            my $bs = perlstring($O::savebackslash) || 'undef';
            $source .= qq(BEGIN { \$/ = $fs; \$\\ = $bs; }\n);
        }

        # I need to do things differently depending on the perl
        # version.
        if ( $] >= 5.008 ) {
            if ( defined &begin_av
                and begin_av->isa('B::AV') )
            {
                $self->todo( $_, 0 ) for begin_av->ARRAY;
            }
            if ( defined &check_av
                and check_av->isa('B::AV') )
            {
                $self->todo( $_, 0 ) for check_av->ARRAY;
            }
            if ( defined &init_av
                and init_av->isa('B::AV') )
            {
                $self->todo( $_, 0 ) for init_av->ARRAY;
            }
            if ( defined &end_av
                and end_av->isa('B::AV') )
            {
                $self->todo( $_, 0 ) for end_av->ARRAY;
            }

            $self->stash_subs;
            $self->{curcv}    = main_cv;
            $self->{curcvlex} = undef;
        }
        else {

            # 5.6.x
            $self->stash_subs('main');
            $self->{curcv} = main_cv;
            $self->walk_sub( main_cv, main_start );
        }

        $source .= join "\n", $self->print_protos;
        @{ $self->{subs_todo} } =
            sort { $a->[0] <=> $b->[0] } @{ $self->{subs_todo} };
        $source .= join "\n", $self->indent( $self->deparse( main_root, 0 ) ),
            "\n"
            unless B::Deparse::null main_root;
        my @text;
        while ( scalar @{ $self->{subs_todo} } ) {
            push @text, $self->next_todo;
        }
        $source .= join "\n", $self->indent( join "", @text ), "\n"
            if @text;

        # Print __DATA__ section, if necessary
        my $laststash =
            defined $self->{curcop}
            ? $self->{curcop}->stash->NAME
            : $self->{curstash};
        {
            ## no critic
            no strict 'refs';
            ## use critic
            if ( defined *{ $laststash . "::DATA" } ) {
                if ( eof *{ $laststash . "::DATA" } ) {

                    # I think this only happens when using B::Deobfuscate
                    # on itself.
                    {
                        local $/ = "__DATA__\n";
                        seek *{ $laststash . "::DATA" }, 0, 0;
                        readline *{ $laststash . "::DATA" };
                    }
                }

                $source .= "__DATA__\n";
                $source .= join '', readline *{ $laststash . "::DATA" };
            }
        }

        my $p    = $self->{ +__PACKAGE__ };
        my %dump = (
            lexicals     => $p->{pad_symbols},
            globals      => $p->{gv_symbols},
            dictionary   => $p->{unknown_dict_file},
            global_regex => $p->{gv_match}
        );

        if ( $p->{output_yaml} ) {
            $p->{output_fh}->print( Dump( \%dump, $source ) );
        }
        else {
            $p->{output_fh}->print($source);
        }

        return;
    };
}

sub padname {
    my $self    = shift @_;
    my $padname = $self->SUPER::padname(@_);

    return $self->rename_pad($padname);
}

sub gv_name {
    my $self    = shift @_;
    my $gv_name = $self->SUPER::gv_name(@_);

    return $self->rename_gv($gv_name);
}

# BEGIN {
#     ## no critic
#     no strict 'refs';
#     for my $sub ( grep defined &$_, keys %B::Deobfuscate:: ) {
#         my $orig = \&$sub;
#         *$sub = sub {
#             print "$sub\n";
#             &$orig;
#         };
#     }
# }

1;

## Local Variables:
## perl-lint-bin: "/home/josh/bin/perl/5.9.4/bin/perl5.9.4"
## eval: (setenv "/home/josh/src/B-Deobfuscate/lib" "PERL5LIB")
## End:
